metricflow/sql/optimizer/optimization_levels.py

```diff
@@ -18,27 +21,49 @@ class SqlQueryOptimizationLevel(Enum):
     O2 = "O2"
     O3 = "O3"
     O4 = "O4"
+    O5 = "O5"
+
 
+@dataclass(frozen=True)
+class SqlQueryGenerationOptionSet:
+    """Defines the different SQL generation optimizers / options that should be used at each level."""
 
-class SqlQueryOptimizerConfiguration:
-    """Defines the different optimizers that should be used at each level."""
+    optimizers: Tuple[SqlQueryPlanOptimizer, ...]
+
+    # Specifies whether CTEs can be used to simplify generated SQL.
+    allow_cte: bool
 
     @staticmethod
-    def optimizers_for_level(
+    def options_for_level(  # noqa: D102
         level: SqlQueryOptimizationLevel, use_column_alias_in_group_by: bool
-    ) -> Sequence[SqlQueryPlanOptimizer]:
-        """Return the optimizers that should be applied (in order) for each level."""
+    ) -> SqlQueryGenerationOptionSet:
+        optimizers: Tuple[SqlQueryPlanOptimizer, ...] = ()
+        allow_cte = False
         if level is SqlQueryOptimizationLevel.O0:
-            return ()
+            pass
         elif level is SqlQueryOptimizationLevel.O1:
-            return (SqlTableAliasSimplifier(),)
+            optimizers = (SqlTableAliasSimplifier(),)
         elif level is SqlQueryOptimizationLevel.O2:
-            return (SqlColumnPrunerOptimizer(), SqlTableAliasSimplifier())
+            optimizers = (SqlColumnPrunerOptimizer(), SqlTableAliasSimplifier())
         elif level is SqlQueryOptimizationLevel.O3:
-            return (SqlColumnPrunerOptimizer(), SqlSubQueryReducer(), SqlTableAliasSimplifier())
+            optimizers = (SqlColumnPrunerOptimizer(), SqlSubQueryReducer(), SqlTableAliasSimplifier())
         elif level is SqlQueryOptimizationLevel.O4:
-            return (
+            optimizers = (
+                SqlColumnPrunerOptimizer(),
+                SqlRewritingSubQueryReducer(use_column_alias_in_group_bys=use_column_alias_in_group_by),
+                SqlTableAliasSimplifier(),
+            )
+        elif level is SqlQueryOptimizationLevel.O5:
+            optimizers = (
                 SqlColumnPrunerOptimizer(),
                 SqlRewritingSubQueryReducer(use_column_alias_in_group_bys=use_column_alias_in_group_by),
                 SqlTableAliasSimplifier(),
             )
+            allow_cte = True
```

Comment:
This one is a little different from the SQL optimizers because it changes how the dataflow plan converts to the SQL plan. The SQL optimizers convert a SQL plan to another SQL plan. Might have been a similar class you saw?